source_code/main_mcu/src/fido2/fido2.c

```diff
@@ -0,0 +1,432 @@
+*   \param  Current count to be updated
+*   \return new count
+*/
+static uint32_t fido2_update_credential_count(uint32_t count)
+{
+    //TBD
+    count++;
+    return count;
+}
+
+/*! \fn     fido2_prompt_user(char const *prompt, bool wait_for_ack, uint32_t timeout_sec)
+*   \brief  Prompt the user and optionally wait for ack from user.
+*   \param  prompt to display
+*   \param  boolean whether to wait for ACK
+*   \param  the amount of time to display message/wait for ack before timeout
+*   \return 0 if accepted or other value if denied/failed
+*/
+static uint32_t fido2_prompt_user(char const *prompt, bool wait_for_ack, uint32_t timeout_sec)
+{
+    uint32_t result = SUCCESS;
+    //TBD
+    //Display request to create cred with name
+    //Wait for ack
+    return result;
+}
+
+/*! \fn     fido2_make_auth_tag(credential_ID_t *cred_ID)
+*   \brief  Make a tag uniquely identifying credential
+*   \param  credential id
+*   \return void
+*/
+static void fido2_make_auth_tag(credential_ID_t *cred_ID)
+{
+    rng_fill_array(cred_ID->tag, FIDO2_TAG_LEN);
+}
+
+/*! \fn     fido2_calc_attestation_signature(uint8_t const *data, int datalen, uint8_t const * client_data_hash, uint8_t * sigbuf)
+*   \brief  Calculate the attestation signature
+*   \param  data to calculate signature over
+*   \param  length of data
+*   \param  hash of client data
+*   \param  out buffer with signature
+*   \return void
+*/
+static void fido2_calc_attestation_signature(uint8_t const *data, int datalen, uint8_t const * client_data_hash, uint8_t * sigbuf)
+{
+    #define HASH_LEN 32
+    uint8_t hashbuf[HASH_LEN];
+
+    _Static_assert(sizeof(hashbuf) >= 32, "Buffer must be 32 bytes or greater");
+
+    logic_encryption_sha256_init();
+    logic_encryption_sha256_update(data, datalen);
+    logic_encryption_sha256_update(client_data_hash, FIDO2_CLIENT_DATA_HASH_LEN);
+    logic_encryption_sha256_final(hashbuf);
+
+    logic_encryption_ecc256_sign(hashbuf, HASH_LEN, sigbuf);
+}
+
+/*! \fn     fido2_cbor_encode_public_key(uint8_t *buf, uint32_t bufLen, ecc256_pub_key const *pub_key)
+*   \brief  CBOR encoder for the public key. All cbor encoding and decoding belongs to aux_mcu.
+*           Unfortunately, to be able to create attestation signature we have to encode the public key
+*           in cbor format so we have to do some cbor encoding on main_mcu also.
+*           Essentiallyt the same as Solo function ctap_add_coseKey() but
+*           removed error checking and pass in buffer instead of the cbor encoder object
+*           in addition to hardcoding algtype.
+*   \param  buffer for encoded output
+*   \param  buffer length
+*   \param  public key
+*   \return encoded public key buffer size
+*/
+static uint32_t fido2_cbor_encode_public_key(uint8_t *buf, uint32_t bufLen, ecc256_pub_key const *pub_key)
+{
+    CborEncoder cose_key;
+    CborEncoder map;
+
+    cbor_encoder_init(&cose_key, buf, bufLen, 0);
```

Comment:
so.... here's an occasion for you to hate me :D.

I see that all the cbor files you add to this main solution are to "only" encode int and byte_string.

Looking at https://en.wikipedia.org/wiki/CBOR#Specification_of_the_CBOR_encoding it seems the encoding logic for ints and bytestrings are fairly simple.... 

So.... how do you feel about removing all the cbor files and do your own cbor encoding in that function only.... as there's no other place where you cbor encode? maybe put the encode primitives in utils.c?

don't hurt me :)